src/instrumentation/do-rpc.ts

```diff
@@ -0,0 +1,249 @@
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions'
+import {  unwrap, wrap } from '../wrap.js'
+import { instrumentEnv } from './env.js'
+import { Initialiser, setConfig, getActiveConfig } from '../config.js'
+import { exportSpans } from './common.js'
+import { DOConstructorTrigger } from '../types.js'
+
+type AlarmFn = DurableObject['alarm']
+type Env = Record<string, unknown>
+type AnyFn = (...args: any[]) => any
+
+const dbSystem = 'Cloudflare DO SQLite'
+
+
+export function getParentContextFromHeaders(headers: Headers): Context {
+	return propagation.extract(api_context.active(), headers, {
+		get(headers, key) {
+			return headers.get(key) || undefined
+		},
+		keys(headers) {
+			return [...headers.keys()]
+		},
+	})
+}
+
+function getParentContextFromRequest(request: Request) {
+	const workerConfig = getActiveConfig()
+
+	if (workerConfig === undefined) {
+		return api_context.active()
+	}
+
+	const acceptTraceContext =
+		typeof workerConfig.handlers.fetch.acceptTraceContext === 'function'
+			? workerConfig.handlers.fetch.acceptTraceContext(request)
+			: (workerConfig.handlers.fetch.acceptTraceContext ?? true)
+	return acceptTraceContext ? getParentContextFromHeaders(request.headers) : api_context.active()
+}
+
+function spanOptions(dbName: string, operation: string, sql?: string): SpanOptions {
+	const attributes: Attributes = {
+		binding_type: 'DO',
+		[SemanticAttributes.DB_NAME]: dbName,
+		[SemanticAttributes.DB_SYSTEM]: dbSystem,
+		[SemanticAttributes.DB_OPERATION]: operation,
+	}
+	if (sql) {
+		attributes[SemanticAttributes.DB_STATEMENT] = sql?.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
+	}
+	return {
+		kind: SpanKind.CLIENT,
+		attributes,
+	}
+}
+
+function metaAttributes(meta: SqlStorageIngestResult): Attributes {
+	return {
+		'db.cf.sql.rows_read': meta.rowsRead,
+		'db.cf.sql.rows_written': meta.rowsWritten
+	}
+}
+
+export function instrumentSqlExec(fn: Function) {
+	const tracer = trace.getTracer('D1')
+	const fnHandler: ProxyHandler<any> = {
+		apply: (target, thisArg, argArray) => {
+			const sql = argArray[0] as string
+			const options = spanOptions('DO SQLite', 'exec', sql)
+			return tracer.startActiveSpan('DO SQL Exec', options, (span) => {
+				try {
+					const result = Reflect.apply(target, thisArg, argArray)
+					span.setAttributes(metaAttributes((result as SqlStorageIngestResult)))
+					span.setStatus({ code: SpanStatusCode.OK })
+					return result
+				} catch (error) {
+					span.recordException(error as Exception)
+					span.setStatus({ code: SpanStatusCode.ERROR })
+					throw error
+				} finally {
+					span.end()
+				}
+			})
+		}
+
+	}
+	return wrap(fn, fnHandler)
+}
+
+export function instrumentSql(sql: SqlStorage) {
+	const sqlHandler: ProxyHandler<SqlStorage> = {
+		get(target, prop, receiver) {
+			const result = Reflect.get(target, prop, receiver)
+			 if (prop === 'exec') {
```

Comment:
There are a few more interesting props. In foundation we have a list of them. e.g., `run`, etc.

Reply:
I'm only handling SQL-Lite methods in this PR

https://developers.cloudflare.com/durable-objects/api/storage-api/